sdks/java/core/src/main/java/org/apache/beam/sdk/coders/BigDecimalCoder.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.coders;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.UTFDataFormatException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+
+/**
+ * A {@link BigDecimalCoder} encodes {@link BigDecimal} in an integer and
+ * a byte array. The integer represents the scale and the byte array
+ * represents a {@link BigInteger}. The integer is in 4 bytes, big-endian.
+ */
+public class BigDecimalCoder extends AtomicCoder<BigDecimal> {
+
+  @JsonCreator
+  public static BigDecimalCoder of() {
+    return INSTANCE;
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+
+  private static final BigDecimalCoder INSTANCE = new BigDecimalCoder();
+
+  private BigDecimalCoder() {}
+
+  @Override
+  public void encode(BigDecimal value, OutputStream outStream, Context context)
+      throws IOException, CoderException {
+    if (value == null) {
+      throw new CoderException("cannot encode a null BigDecimal");
+    }
+
+    byte[] bigIntBytes = value.unscaledValue().toByteArray();
+
+    DataOutputStream dataOutputStream = new DataOutputStream(outStream);
+    dataOutputStream.writeInt(value.scale());
+    dataOutputStream.writeInt(bigIntBytes.length);
+    dataOutputStream.write(bigIntBytes);
+  }
+
+  @Override
+  public BigDecimal decode(InputStream inStream, Context context)
+      throws IOException, CoderException {
+    try {
+      DataInputStream dataInputStream = new DataInputStream(inStream);
+      int scale = dataInputStream.readInt();
+      int bigIntBytesSize = dataInputStream.readInt();
+
+      byte[] bigIntBytes = new byte[bigIntBytesSize];
+      dataInputStream.readFully(bigIntBytes);
+
+      BigInteger bigInteger = new BigInteger(bigIntBytes);
+      BigDecimal bigDecimal = new BigDecimal(bigInteger, scale);
+
+      return bigDecimal;
+    } catch (EOFException | UTFDataFormatException exn) {
```

Comment:
I think I see what you're saying now. I shouldn't be doing a try/catch there. Will remove it.

I followed `BigEndianIntegerCoder` as an example. It does a try/catch for the same exceptions. A quick look found some other primitive coders doing the same thing. I don't know if those are an issue too.